### DISCUSSION TITLE
Check if old password matches

### DISCUSSION THREAD

**Discussion by Paul6552** [keycloak/keycloak#23696] [Q&A]
Can I warm up this question:
[Check if old password matches](https://stackoverflow.com/questions/64726445/keycloak-check-if-old-password-matches-if-so-change-to-another-one)

Two years have now passed, perhaps there is now a solution?

Short:
How can i check if the entered password matches the old password in java?

**Comment by kwame-mintah** [comment]
@Paul6552 Any reason why enforcing a password policy wouldn't be sufficient?
https://www.keycloak.org/docs/latest/server_admin/index.html#not-recently-used

**Reply by Paul6552** [reply]
I am sorry, but I do not really understand what you mean. Why would a password policy solve this problem?

I mean something like that:
![image](https://github.com/keycloak/keycloak/assets/25934591/a598f1fa-867b-48b3-a230-c08025ee6550)

Before you can change your password, I would like to check the old password first.

**Reply by kwame-mintah** [reply]
@Paul6552, sorry if I have misunderstood your question. But my understanding is that, you want to check the password the user has entered, before updating the password in Keycloak with Java?

If you enable the password policy `not-recently-used`, you don't have to check yourself. Because having the policy enabled means Keycloak will check and return an error and all you have to do is pick up on the this error message / exception returned:

https://github.com/keycloak/keycloak/blob/main/themes/src/main/resources/theme/base/account/messages/messages_en.properties#L231

**Reply by Paul6552** [reply]
no no.

I would like to check the current password from the user. Like my picture from above, I would like to change to a new password, but before I confirm this change I would like to check if the current password is correct. So It is only about checking his current / old password before changing to a new one.

**Reply by kwame-mintah** [reply]
Right, so you just want to check the users' password before you update it. And don't want them to have entered an old password? I think the snippet below might be able to help:

```java
UserModel userModel = <USER>?
List<CredentialModel> credentials = userModel.credentialsManager().getStoredCredentialsByTypeStream("").toList();
```

Java docs for `getStoredCredentialsByTypeStream()` here:
https://www.keycloak.org/docs-api/22.0.4/javadocs/org/keycloak/models/SubjectCredentialManager.html#getStoredCredentialsByTypeStream(java.lang.String)

I am unable to help any further, as I don't know what the credential will be stored under in Keycloak. Could be `"keycloak-password"`, but don't have time to investigate any further.

**Comment by buhtr** [comment]
I've created a project https://github.com/buhtr/keycloak-check-password/tree/main which demonstrates how this task can be solved. If you have any questions, please feel free to let me know
